### PR TITLE
TeamScan for CUDA, Pthreads, OpenMPTarget, HIP

### PR DESCRIPTION
**Pull request by jrmadsen**
- implemented team-level parallel scan in CUDA

### CODE REVIEW

**Comment by jrmadsen** [comment]
- Addresses #3197 
- Addresses #45

**Comment by crtrott** [comment]
impl scatch
Threaded:
```c++
parallel_scan(..., N,...) {
  my_sum;
  for(int i = N/team_size*team_rank; ... ) {

     f(i,my_sum,false);
  }
  offset = team.team_scan(my_sum);
  my_sum = 0;
  for(int i = N/team_size*team_rank; ... ) {
     f(i,my_sum+offset,false);
  }
}
```
Cuda:
```
parallel_scan(..., N,...) {
  my_sum;
  offset = 0;
  for(int chunks = N/team_size) {
     f(chunks*team_size+team_rank,my_sum,false);
    local_offset = team.team_scan(my_sum);
     f(chunks*team_size+team_rank,my_sum+local_offset+offset,true);
     if(team_rank()==team_size-1)
     offset += ...
  }
}
```

**Comment by jrmadsen** [comment]
@crtrott I ended up with a slightly different implementation that you recommended. This appears to work as long as the team-size is a power of 2, otherwise the `team_scan` throws an error.

**Comment by dalg24** [comment]
Retest this please

**Review by crtrott — CHANGES_REQUESTED** [review]
We met today and discussed correctness issues. Jonathan is working on this.

**Comment by jrmadsen** [comment]
@dalg24 @masterleinad Do either of y'all know why the Jenkins build failed? I searched for all instances of the word "error" and "fail" in the Jenkins log and basically the only "failure" was "script exited with error code 2" but the build and tests appear to be fine

**Comment by masterleinad** [comment]
```
/var/jenkins/workspace/Kokkos/core/unit_test/TestTeamScan.hpp:66:16: error: unused variable 'teamSize' [clang-diagnostic-unused-variable]
          auto teamSize   = team.team_size();
               ^
/var/jenkins/workspace/Kokkos/core/unit_test/TestTeamScan.hpp:87:5: note: in instantiation of member function 'Test::TestTeamScan<Kokkos::Serial, short>::operator()' requested here
    (*this)(M, N, a_d, a_r);
    ^
```

**Comment by masterleinad** [comment]
```
/var/jenkins/workspace/Kokkos/install/include/Cuda/Kokkos_Cuda_Parallel.hpp(680): error: calling a constexpr __host__ function("operator()") from a __device__ function("exec_team") is not allowed. The experimental flag '--expt-relaxed-constexpr' can be used to allow this.
          detected during:
            instantiation of "std::enable_if<std::is_same<TagType, void>::value, void>::type Kokkos::Impl::ParallelFor<FunctorType, Kokkos::TeamPolicy<Properties...>, Kokkos::Cuda>::exec_team<TagType>(const Kokkos::Impl::ParallelFor<FunctorType, Kokkos::TeamPolicy<Properties...>, Kokkos::Cuda>::Member &) const [with FunctorType=lambda [](const Kokkos::Impl::CudaTeamMember &)->void, Properties=<Kokkos::Cuda>, TagType=void]" 
(731): here
            instantiation of "void Kokkos::Impl::ParallelFor<FunctorType, Kokkos::TeamPolicy<Properties...>, Kokkos::Cuda>::operator()() const [with FunctorType=lambda [](const Kokkos::Impl::CudaTeamMember &)->void, Properties=<Kokkos::Cuda>]" 
/var/jenkins/workspace/Kokkos/install/include/Cuda/Kokkos_Cuda_KernelLaunch.hpp(121): here
            instantiation of "void Kokkos::Impl::cuda_parallel_launch_local_memory(DriverType) [with DriverType=Kokkos::Impl::ParallelFor<lambda [](const Kokkos::Impl::CudaTeamMember &)->void, Kokkos::TeamPolicy<Kokkos::Cuda>, Kokkos::Cuda>]" 
/var/jenkins/workspace/Kokkos/install/include/Cuda/Kokkos_Cuda_KernelLaunch.hpp(319): here
            instantiation of "std::decay_t<decltype((<expression>))> Kokkos::Impl::CudaParallelLaunchKernelFunc<DriverType, Kokkos::LaunchBounds<0U, 0U>, Kokkos::Impl::Experimental::CudaLaunchMechanism::LocalMemory>::get_kernel_func() [with DriverType=Kokkos::Impl::ParallelFor<lambda [](const Kokkos::Impl::CudaTeamMember &)->void, Kokkos::TeamPolicy<Kokkos::Cuda>, Kokkos::Cuda>]" 
/var/jenkins/workspace/Kokkos/install/include/Cuda/Kokkos_Cuda_KernelLaunch.hpp(646): here
            instantiation of "cudaFuncAttributes Kokkos::Impl::CudaParallelLaunchImpl<DriverType, Kokkos::LaunchBounds<MaxThreadsPerBlock, MinBlocksPerSM>, LaunchMechanism>::get_cuda_func_attributes() [with DriverType=Kokkos::Impl::ParallelFor<lambda [](const Kokkos::Impl::CudaTeamMember &)->void, Kokkos::TeamPolicy<Kokkos::Cuda>, Kokkos::Cuda>, MaxThreadsPerBlock=0U, MinBlocksPerSM=0U, LaunchMechanism=Kokkos::Impl::Experimental::CudaLaunchMechanism::LocalMemory]" 
(764): here
            instantiation of "Kokkos::Impl::ParallelFor<FunctorType, Kokkos::TeamPolicy<Properties...>, Kokkos::Cuda>::ParallelFor(const FunctorType &, const Kokkos::Impl::ParallelFor<FunctorType, Kokkos::TeamPolicy<Properties...>, Kokkos::Cuda>::Policy &) [with FunctorType=lambda [](const Kokkos::Impl::CudaTeamMember &)->void, Properties=<Kokkos::Cuda>]" 
/var/jenkins/workspace/Kokkos/install/include/Kokkos_Parallel.hpp(168): here
            instantiation of "void Kokkos::parallel_for(const ExecPolicy &, const FunctorType &, const std::__cxx11::string &, std::enable_if<Kokkos::is_execution_policy<ExecPolicy>::value, void>::type *) [with ExecPolicy=Kokkos::TeamPolicy<Kokkos::Cuda>, FunctorType=lambda [](const Kokkos::Impl::CudaTeamMember &)->void]" 
/var/jenkins/workspace/Kokkos/core/unit_test/TestTeamScan.hpp(80): here
            instantiation of "void Test::TestTeamScan<Device, DataType>::operator()(int32_t, int32_t, Test::TestTeamScan<Device, DataType>::view_type, Test::TestTeamScan<Device, DataType>::view_type) const [with Device=Kokkos::Cuda, DataType=int16_t]" 
/var/jenkins/workspace/Kokkos/core/unit_test/TestTeamScan.hpp(87): here
```

**Comment by masterleinad** [comment]
```
/var/jenkins/workspace/Kokkos/core/unit_test/TestTeamScan.hpp:125:57: error: typedef 'using TEST_POLICY = class Kokkos::TeamPolicy<Kokkos::OpenMP>' locally defined but not used [-Werror=unused-local-typedefs]
   using TEST_POLICY = Kokkos::TeamPolicy<TEST_EXECSPACE>;
                                                         ^
```

**Comment by jrmadsen** [comment]
@crtrott @dalg24 So everything is passing here except for some SYCL stuff, which I didn't touch but it looks like the new tests instantiate something that is incomplete in the SYCL backend. How do we proceed here?

File: core/unit_test/CMakeLists.txt [hunk 1]
```diff
@@ -124,6 +124,7 @@ foreach(Tag Threads;Serial;OpenMP;Cuda;HPX;OpenMPTarget;HIP)
     foreach(Name
       TeamBasic
       TeamReductionScan
+      TeamScan
```
Comment (masterleinad) [inline core/unit_test/CMakeLists.txt:127]: Just remove this test for `SYCL`. There is a list further down for that.

**Review by crtrott — APPROVED** [review]
Minus the "exclude the test in the exclude list for SYCL" this looks good!

File: core/unit_test/CMakeLists.txt (hunk 1 above)
Comment (jrmadsen) [reply core/unit_test/CMakeLists.txt:130]: Ah, I had a lot of trouble finding it but just needed to rebase

**Review by e10harvey — APPROVED** [review]

**Review by Rombur — APPROVED** [review]
The HIP backend looks good

File: core/unit_test/TestTeamScan.hpp [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+          std::stringstream ss, idx;
+          idx << "(" << i << ", " << j << ") = ";
+          ss << "a_d" << idx.str() << a_i(i, j);
+          ss << ", a_r" << idx.str() << a_o(i, j);
+          return ss.str();
+        };
+        if (std::is_integral<value_type>::value) {
+          ASSERT_EQ(_scan_real, _scan_calc) << _get_mesg();
+        } else {
+          _epsilon += std::numeric_limits<value_type>::round_error();
```
Comment (dalg24) [inline core/unit_test/TestTeamScan.hpp:119]: `std::numeric_limits<FLoatingPoint>::round_error()` is `0.5`

**Review by dalg24 — CHANGES_REQUESTED** [review]
Please clarify the tolerance with floating point numbers

File: core/src/Cuda/Kokkos_Cuda_Team.hpp [hunk 3]
```diff
@@ -935,6 +935,52 @@ KOKKOS_INLINE_FUNCTION
...
+  const auto nchunk    = (end - start) / team_size;
+  value_type accum     = 0;
+  value_type offset    = 0;
+
+  for (iType i = 0; i < nchunk + 1; ++i) {
+    auto ii                 = i * team_size + team_rank;
+    value_type local_offset = 0;
+    if (ii < loop_bounds.end) lambda(ii, local_offset, false);
+    local_offset = member.team_scan(local_offset);
+    auto val     = accum + offset + local_offset;
```
Comment (dalg24) [inline core/src/Cuda/Kokkos_Cuda_Team.hpp:971]: Why do you have `accum` at all?  It is defined to be zero and never updated.

File: core/src/Cuda/Kokkos_Cuda_Team.hpp [hunk 4]
```diff
@@ -935,6 +935,52 @@ KOKKOS_INLINE_FUNCTION
...
+ *  The range [0..N) is mapped to all vector lanes in the
+ *  thread and a scan operation is performed.
+ *  The last call to closure has final == true.
+ */
+template <typename iType, typename FunctorType>
+KOKKOS_INLINE_FUNCTION void parallel_scan(
+    const Impl::TeamThreadRangeBoundariesStruct<iType, Impl::CudaTeamMember>&
+        loop_bounds,
+    const FunctorType& lambda) {
+#ifdef __CUDA_ARCH__
```
[the remark this answers is not in the document]
Comment (jrmadsen) [reply core/src/Cuda/Kokkos_Cuda_Team.hpp:951]: I don't. Left over from when I had CUDA specific stuff

File: core/unit_test/TestTeamScan.hpp [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+  TestTeamScan<TEST_EXECSPACE, uint16_t>{}(139, 64);
+  TestTeamScan<TEST_EXECSPACE, int32_t>{}(163, 128);
+  TestTeamScan<TEST_EXECSPACE, int64_t>{}(2156, 512);
+  TestTeamScan<TEST_EXECSPACE, uint64_t>{}(1234, 1024);
+  TestTeamScan<TEST_EXECSPACE, float>{}(108, 16);
+  TestTeamScan<TEST_EXECSPACE, float>{}(152, 80);
+  TestTeamScan<TEST_EXECSPACE, double>{}(34, 32);
+  TestTeamScan<TEST_EXECSPACE, double>{}(956, 128);
+  TestTeamScan<TEST_EXECSPACE, double>{}(2596, 512);
+  TEST_EXECSPACE().fence();
```
[the remark this answers is not in the document]
Comment (jrmadsen) [reply core/unit_test/TestTeamScan.hpp:138]: Left over from another test

File: core/unit_test/TestTeamScan.hpp [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+  KOKKOS_FUNCTION
+  void operator()(const member_type& team) const {
+    auto leagueRank = team.league_rank();
+
+    auto beg = 0;
+    auto end = N;
+
+    Kokkos::parallel_for(
+        Kokkos::TeamThreadRange(team, beg, end),
+        [&](const int i) { a_d(leagueRank, i) += leagueRank * N + i; });
```
[the remark this answers is not in the document]
Comment (jrmadsen) [reply core/unit_test/TestTeamScan.hpp:75]: Left over from original implementation

File: core/unit_test/TestTeamScan.hpp [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+
+    auto a_i = Kokkos::create_mirror_view(a_d);
+    auto a_o = Kokkos::create_mirror_view(a_r);
+    Kokkos::deep_copy(a_i, a_d);
+    Kokkos::deep_copy(a_o, a_r);
+
+    EXPECT_EQ(a_i.extent(0), M);
+    EXPECT_EQ(a_o.extent(0), M);
+    EXPECT_EQ(a_i.extent(1), N);
+    EXPECT_EQ(a_o.extent(1), N);
```
[the remark this answers is not in the document]
Comment (jrmadsen) [reply core/unit_test/TestTeamScan.hpp:100]: Fair enough

File: core/unit_test/TestTeamScan.hpp (hunk 2 above)
Comment (jrmadsen) [reply core/unit_test/TestTeamScan.hpp]: My bad, I meant `std::numeric_limits<value_type>::epsilon();`

File: core/unit_test/TestTeamScan.hpp [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+
+    auto beg = 0;
+    auto end = N;
+
+    Kokkos::parallel_for(
+        Kokkos::TeamThreadRange(team, beg, end),
+        [&](const int i) { a_d(leagueRank, i) = leagueRank * N + i; });
+
+    Kokkos::parallel_scan(Kokkos::TeamThreadRange(team, beg, end),
+                          [&](int i, int32_t& val, const bool final) {
```
Comment (dalg24) [inline core/unit_test/TestTeamScan.hpp:78]: Did you intend `val` to be `int32_t` or should it be `DataType`?

File: core/unit_test/TestTeamScan.hpp [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+  TestTeamScan<TEST_EXECSPACE, int16_t>{}(99, 32);
+  TestTeamScan<TEST_EXECSPACE, uint16_t>{}(139, 64);
+  TestTeamScan<TEST_EXECSPACE, int32_t>{}(163, 128);
+  TestTeamScan<TEST_EXECSPACE, int64_t>{}(2156, 512);
+  TestTeamScan<TEST_EXECSPACE, uint64_t>{}(1234, 1024);
+  TestTeamScan<TEST_EXECSPACE, float>{}(108, 16);
+  TestTeamScan<TEST_EXECSPACE, float>{}(152, 80);
+  TestTeamScan<TEST_EXECSPACE, double>{}(34, 32);
+  TestTeamScan<TEST_EXECSPACE, double>{}(956, 128);
+  TestTeamScan<TEST_EXECSPACE, double>{}(2596, 512);
```
Comment (dalg24) [inline core/unit_test/TestTeamScan.hpp:138]: Please add a few tests where the second argument is not a multiple of 16 and one where it is zero.

Comment (jrmadsen) [reply core/unit_test/TestTeamScan.hpp]: Done.

File: core/unit_test/TestTeamScan.hpp (hunk 8 above)
Comment (jrmadsen) [reply core/unit_test/TestTeamScan.hpp]: Good catch. Thats very Interesting that the lambda compiled for all the backends since it takes a reference...

**Review by dalg24 — APPROVED** [review]

**Comment by crtrott** [comment]
Retest this please.

**Comment by crtrott** [comment]
So intereseting question: should we just merge? Windows likely failing because of the MSVC stuff, Jenkins its the AMD node, and travis is one timeout ...

**Comment by crtrott** [comment]
Also: Jonathan you want to rewrite history or should I squash commit?

**Comment by Char-Aznable** [comment]
Hi @ jrmadsen @crtrott  , I have code using something like `parallel_scan(TeamThreadRange(team, n), [&](const int i, double& udpate, const bool isFinal) { ... });` and it aborts at the line https://github.com/kokkos/kokkos/blob/e4831441fb213bd29485327b02cf1b54149a58a7/core/src/Cuda/Kokkos_Cuda_ReduceScan.hpp#L698-L700 and 
```c++
(cuda-gdb) p blockDim.y
$4 = 224
(cuda-gdb) p blockDim.y - 1
$5 = 223

```
any idea what went wrong here? Does the code in this PR require the item counts `n` to be the same as number of threads in the team because it seems to be calling `team_member.team_scan()` judging from the stack trace?

**Comment by Char-Aznable** [comment]
Judging from the unit test cases, I guess this implementation only works if the work item counts to TeamThreadRange is power of 2?

**Comment by jrmadsen** [comment]
> Judging from the unit test cases, I guess this implementation only works if the work item counts to TeamThreadRange is power of 2?

Yes

**Comment by Char-Aznable** [comment]
Would it be difficult to support non-power-of-2 work counts? I can help implementing it with the CUDA backend at least. I have a lot of small loops of a few hundreds work counts and it would be a big hit in performance if forced to use power-of-2 loop, especially on the host because I need to convert all the host side to be power of 2 to be portable

**Comment by masterleinad** [comment]
We will discuss this but in general, contributions are very welcome.

**Comment by masterleinad** [comment]
@Char-Aznable We decided that we are interested in making non-power-2 team sizes work, see https://github.com/kokkos/kokkos/issues/4146. Any help in implementing that is very welcome!

**Comment by Char-Aznable** [comment]
Great! I'll take a look at the code and see what I can do
